keras/src/quantizers/gptq.py

```diff
@@ -0,0 +1,252 @@
+            except ValueError:
+                raise TypeError(
+                    f"Could not determine hidden dimension from shape {shape}"
+                )
+
+            if d_model_dim_index == 0:  # QKV projection case
+                in_features, heads, head_dim = shape
+                self.rows, self.columns = in_features, heads * head_dim
+            elif d_model_dim_index in [1, 2]:  # Attention Output case
+                heads, head_dim, out_features = shape
+                self.rows, self.columns = heads * head_dim, out_features
+
+            # Create a temporary object that holds a reshaped
+            # 2D version of the kernel.
+            self.layer = type(
+                "temp",
+                (object,),
+                {
+                    "kernel": ops.reshape(
+                        layer.kernel, (self.rows, self.columns)
+                    ),
+                    "bias": layer.bias,
+                },
+            )()
+
+        else:
+            # Raise an error if the layer is not supported.
+            raise TypeError(
+                f"Unsupported layer type or kernel shape for GPTQ: "
+                f"{type(layer)} with kernel ndim {layer.kernel.ndim}"
+            )
+        self.H = ops.zeros((self.rows, self.rows), dtype="float32")
+
+    def update_hessian_with_batch(self, inp):
+        """
+        Updates the running average of the Hessian matrix with a new batch.
+
+        This method computes the Hessian matrix for a given batch of input
+        activations and updates the accumulated Hessian (`self.H`) using a
+        numerically stable running average. This allows the Hessian to be
+        computed over a large dataset without loading all samples into memory
+        at once.
+
+        The input tensor is first reshaped into a 2D matrix [num_samples,
+        num_features] before the Hessian is calculated.
+
+        Args:
+            inp: A 2D or higher-dimensional tensor of input activations from a
+                calibration batch.
+
+        Raises:
+            ValueError: If the feature dimension of the input tensor `inp` does
+                not match the dimensions of the pre-initialized Hessian matrix
+                `self.H`.
+        """
+        if len(inp.shape) > 2:
+            inp = ops.reshape(inp, (-1, inp.shape[-1]))
+        inp = ops.cast(inp, "float32")
+
+        if self.H.shape[0] != inp.shape[-1]:
+            raise ValueError(
+                f"Hessian dimensions ({self.H.shape[0]}) do not"
+                "match input features ({inp.shape[-1]})."
+            )
+
+        current_H = 2 * ops.matmul(ops.transpose(inp), inp)
+
+        if self.nsamples == 0:
+            self.H = current_H
+        else:
+            self.H = self.H * (self.nsamples / (self.nsamples + inp.shape[0]))
+            self.H += current_H * (
+                inp.shape[0] / (self.nsamples + inp.shape[0])
+            )
```

Comment:
Let's use the built in keras ops instead of direct `/`, `*`, etc. This ensures the computations happen through optimized backend-specific paths.